synthesizer/src/coinbase_puzzle/mod.rs

```diff
@@ -208,14 +210,10 @@ impl<N: Network> CoinbasePuzzle<N> {
             None => bail!("Missing the accumulator challenge point"),
         };
 
-        // Construct the provers polynomial.
+        // Accumulate the prover polynomial.
         let accumulated_prover_polynomial = cfg_into_iter!(prover_polynomials)
             .zip_eq(challenges)
-            .fold(DensePolynomial::zero, |mut accumulator, (mut prover_polynomial, challenge)| {
-                prover_polynomial *= challenge;
-                accumulator += &prover_polynomial;
-                accumulator
-            })
+            .map(|(prover_polynomial, challenge)| prover_polynomial * challenge)
             .sum::<DensePolynomial<_>>();
```

Comment:
Since the only diff between `rayon::ParallelIterator::fold` and `std::Iterator::fold` is in the initialization value (the former takes a closure while the latter directly takes a value), a suggestion to deduplicate logic would be to do the following:

```rust
#[cfg(serial)]
let init = DensePolynomial::zero();
#[cfg(not(feature = "serial")]
let init = DensePolynomial::zero;

let thing = cfg_iter!(...).fold(init, stuff);
```

We want to preserve the old behaviour as it reduces memory consumption

Reply:
We may want to just do the "dumb" approach of having a full separation of the `serial` and non-`serial` logic, because the 
`rayon::ParallelIterator::fold` and `std::Iterator::fold` differs in the initalizer and final `sum` operation. 

The parallel iterator requires the `sum` operation because the objects are [disjointly folded](https://docs.rs/rayon/latest/rayon/iter/trait.ParallelIterator.html#method.fold) into breakpoints, whereas `std::Iterator::fold` does not require a `sum` operation.


Example:
```
        #[cfg(feature = "serial")]
        let mut accumulator_evaluation = cfg_iter!(prover_polynomials).zip_eq(&challenge_points).fold(
            <N::PairingCurve as PairingEngine>::fr::zero(),
            |accumulator, (prover_polynomial, challenge_point)| {
                accumulator + (prover_polynomial.evaluate(accumulator_point) * challenge_point)
            },
        );

        #[cfg(not(feature = "serial"))]
        let mut accumulator_evaluation = cfg_iter!(prover_polynomials)
            .zip_eq(&challenge_points)
            .fold(<N::PairingCurve as PairingEngine>::fr::zero, |accumulator, (prover_polynomial, challenge_point)| {
                accumulator + (prover_polynomial.evaluate(accumulator_point) * challenge_point)
            })
            .sum();
```

This is more verbose, however more clear because it doesn't require an additional `cfg` flag for the final sum operation:
``` 
#[cfg(not(feature = "serial"))]
let accumulator_evaluation = accumulator_evaluation.sum::<DensePolynomial<_>>();
```

Reply:
I pushed up a macro for the `fold` operation as Pratyush suggested, but I think there's an argument for the dumb way. Lots of rust macros often save a few lines but they also create levels of indirection when trying to understand the code. I'd personally pick the "dumb" approach here for readability, but I used a macro as suggested!